docs: restructure simulation/emulation and expand config logic chapter (#953)
FPGA-Research/FABulous: docs/source/user_guide/simulation/configuration.md

(fabric-configuration)=
# Fabric configuration

FABulous fabrics can be loaded with a bitstream through one of three
configuration methods. For how to run a simulation, see

> [Simulation setup](simulation.md).

Reviewer: ```suggestion
FABulous fabrics offer a 32-bit wide configuration port which can either be driven directly by e.g. a CPU or bus or by the two currently implemented configuration adapters, allowing the transmission over UART and a custom bitbang protocol.
```

:::{note}
The FABulous testbench currently drives configuration only through the
**parallel port (Mode 1)**. Serial (Mode 0) and Bitbang are supported by the
fabric RTL but are not yet exercised by the generated testbench.

> :::

Reviewer: If we (rightfully so!) move the config logic description to a separate chapter, we then should not instead move the simulation information into this new chapter, but rather keep it truly separated.

> ## Parallel (Mode 1) — default in testbench

Reviewer: ```suggestion
## Parallel Port
```

Configuration data is written directly into the fabric through a 32-bit
parallel port, rather than shifted in bit-serially. This is the fastest of
the three methods and is the only one currently exercised by the FABulous

> testbench.

Reviewer: ```suggestion
parallel port, rather than shifted in bit-serially. This is the main configuration interface which is also used by the UART and bitbang configuration adapters.
```

- **Signals:** `SelfWriteData` (32-bit data bus), `SelfWriteStrobe` (write strobe)
- **Bus width:** 32 bits — each transfer loads 4 bytes of the bitstream
- **Timing:** for each word, data is held for 2 clock cycles, then
  `SelfWriteStrobe` is pulsed high for one clock cycle to latch it in,
  followed by 2 more idle cycles before the next word

> - **Transfer count:** `MAX_BITBYTES / 4` word writes to load the full bitstream

Reviewer: I think this information is extracted from the testbench? Especially the transfer count does not necessarily have to be mentioned here. The timing can be mentioned here, but it needs to be double checked if this is a generic description of the interface or rather just distilled from the testbench (e.g. write something like "data needs to be valid for x clock cycles before the `SelfWriteStrobe`).

> ## Serial (Mode 0)

Reviewer: ```suggestion
## Configuration Adapters

To allow configuration from outside of a chip, currently two configuration adapters are implemented, allowing to upload a bitstream using a UART or a custom bitstream protocol.

###  UART
```

I would suggest that we add an own section for UART and bitbang, since they are technically only adapters to the config interface.

Configuration is sent to the fabric byte-by-byte over a UART link on `Rx`,
decoded, and assembled into 32-bit words that feed the same write interface

> used by Parallel mode.

Reviewer: ```suggestion
The configuration is sent to the fabric bit-by-bit over a UARTlink on `Rx (since the adapter does not transmit anything, it's technically just a "UAR")`,
decoded, and assembled into 32-bit words that feed the configuration port.
```

- **Signals:** `Rx` (UART input), `WriteData` (32-bit, shared with parallel
  write path), `WriteStrobe`, `Command` (decoded command byte), `ComActive`,
  `ReceiveLED`
- **Frame format:** standard UART frame — 1 start bit, 8 data bits, 1 stop bit
- **Baud rate:** set by `ComRate = f_CLK / Baudrate` (default `217`, e.g. 25 MHz / 115200 baud)
- **Encoding modes:** `auto`, `hex` (2 ASCII hex chars per byte), or `bin`
  (raw byte), selected by the `Mode` parameter or auto-detected from the
  command byte
- **Framing:** each transfer begins with a fixed ID header and a command
  byte before data bytes are accepted
- **Word assembly:** 4 received bytes are packed into one `WriteData` word,

>   then `WriteStrobe` pulses once per word — matching the Parallel-mode timing

Reviewer: ```suggestion
  then `WriteStrobe` pulses once per word, as the configuration interface expects
```

- **Integrity check:** a running checksum is accumulated over the data and
  validated against an expected checksum value
- **Timeout:** an inactivity timeout resets the receiver to idle if no data
  arrives mid-transfer

> ## Bitbang configuration port (To be supported in the testbench)

Reviewer: ```suggestion
### Bitbang
```

> We have produced a quick asynchronous serial configuration port interface that is ideal for microcontroller configuration. It uses the original CPU interface that we have in our TSMC chip. The idea of the protocol is as follows:

Reviewer: ```suggestion
The bitbang adapter offers a quick asynchronous serial configuration port interface that is ideal for configuring the fabric via a microcontroller. The idea of the protocol is as follows:
```

:::{figure} ./figs/bitbang1.*
:align: center
:alt: Bitbang description
:::

> We drive s_clk and s_data. On each rising edge of s_clock, we sample data and on the falling edge, we sample control.

Reviewer 2: ```suggestion
We drive `s_clk` and `s_data`. On each rising edge of s_clock, we sample data and on the falling edge, we sample control.
```

> We drive `s_clk` and `s_data`. On each rising edge of s_clock, we sample data and on the falling edge, we sample control.

Reviewer: ```suggestion
The protocol uses the two signals `s_clk` and `s_data`. On each rising edge of `s_clk`, we sample data and on the falling edge, we sample control.
```
